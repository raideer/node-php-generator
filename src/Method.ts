import { asFunction } from "mixins/asFunction";
import { withAttributes } from "mixins/withAttributes";
import { withComments } from "mixins/withComments";
import { withName } from "mixins/withName";
import { withVisibility } from "mixins/withVisibility";
import { PromotedParameter } from "./PromotedParameter";

export class Method extends asFunction(withVisibility(withAttributes(withComments(withName(class {}))))) {
  public static: boolean = false;
  public final: boolean = false;
  public abstract: boolean = false;

  setStatic(isStatic: boolean): this {
    this.static = isStatic;

    return this;
  }

  isStatic(): boolean {
    return this.static;
  }

  setFinal(final: boolean): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setAbstract(abstract: boolean): this {
    this.abstract = abstract;

    return this;
  }

  isAbstract(): boolean {
    return this.abstract;
  }

  addPromotedParameter(name: string, defaultValue: string | null | undefined): this {
    const parameter = new PromotedParameter();
    parameter.setName(name);

    if (defaultValue !== undefined) {
      parameter.setDefaultValue(defaultValue);
    }

    this.parameters[name] = parameter;

    return this;
  }
}
