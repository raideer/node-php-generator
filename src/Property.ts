import { CanValidate } from './Api/CanValidate';
import { HasAttributes } from './Api/HasAttributes';
import { HasComment } from './Api/HasComment';
import { HasName } from './Api/HasName';
import { HasVisibility } from './Api/HasVisibility';
import { Attribute, AttributeArguments } from './Attribute';
import { PropertyValue, Visibility } from './types';

export class Property
  implements HasName, HasVisibility, HasComment, HasAttributes, CanValidate
{
  private name: string;
  private visibility: Visibility | undefined;
  private comment: string | undefined;
  private attributes: Attribute[] = [];

  private value: PropertyValue;
  private static: boolean = false;
  private type: string | undefined;
  private nullable: boolean = false;
  private initialized: boolean = false;
  private readOnly: boolean = false;

  constructor(name: string) {
    this.name = name;
  }

  public getName(): string {
    return this.name;
  }

  public setName(name: string): this {
    this.name = name;
    return this;
  }

  public getVisibility(): Visibility | undefined {
    return this.visibility;
  }

  public setVisibility(visibility: Visibility | undefined): this {
    this.visibility = visibility;
    return this;
  }

  public setPublic(): this {
    this.visibility = 'public';
    return this;
  }

  public isPublic(): boolean {
    return this.visibility === 'public';
  }

  public setPrivate(): this {
    this.visibility = 'private';
    return this;
  }

  public isPrivate(): boolean {
    return this.visibility === 'private';
  }

  public setProtected(): this {
    this.visibility = 'protected';
    return this;
  }

  public isProtected(): boolean {
    return this.visibility === 'protected';
  }

  public getComment(): string | undefined {
    return this.comment;
  }

  public setComment(comment: string | undefined): this {
    this.comment = comment;
    return this;
  }

  public addComment(comment: string): this {
    this.comment = (this.comment ? this.comment + '\n' : '') + comment;
    return this;
  }

  public removeComment(): this {
    this.comment = undefined;
    return this;
  }

  public getAttributes(): Attribute[] {
    return this.attributes;
  }

  public addAttribute(name: string, args: AttributeArguments = {}): this {
    this.attributes.push(new Attribute(name, args));
    return this;
  }

  public setAttributes(attributes: Attribute[]): this {
    this.attributes = attributes;
    return this;
  }

  public setValue(value: PropertyValue): this {
    this.value = value;
    this.initialized = true;
    return this;
  }

  public getValue(): PropertyValue {
    return this.value;
  }

  public setStatic(stat: boolean = true): this {
    this.static = stat;
    return this;
  }

  public isStatic(): boolean {
    return this.static;
  }

  public setType(type: string | undefined): this {
    if (type?.startsWith('?')) {
      this.nullable = true;
      type = type.slice(1);
    }

    this.type = type;
    return this;
  }

  public getType(): string | undefined {
    return this.type;
  }

  public isNullable(): boolean {
    return this.nullable || (this.initialized && this.value === null);
  }

  public setNullable(nullable: boolean = true): this {
    this.nullable = nullable;
    return this;
  }

  public setInitialized(initialized: boolean = true): this {
    this.initialized = initialized;
    return this;
  }

  public isInitialized(): boolean {
    return this.initialized || this.value !== undefined;
  }

  public setReadOnly(readOnly: boolean = true): this {
    this.readOnly = readOnly;
    return this;
  }

  public isReadOnly(): boolean {
    return this.readOnly;
  }

  public validate(): void {
    if (this.readOnly && !this.type) {
      throw new Error(
        'Property cannot be read-only if it does not have a type'
      );
    }
  }

  public cloneWithName(name: string): this {
    const property = structuredClone(this);
    property.name = name;
    return property;
  }
}
