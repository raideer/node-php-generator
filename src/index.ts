export { Attribute } from './Attribute';
export { ClassType } from './ClassType';
export { Closure } from './Closure';
export { Constant } from './Constant';
export { EnumCase } from './EnumCase';
export { EnumType } from './EnumType';
export { GlobalFunction } from './GlobalFunction';
export { InterfaceType } from './InterfaceType';
export { Literal } from './Literal';
export { Method } from './Method';
export { Parameter } from './Parameter';
export { PhpFile } from './PhpFile';
export { PhpNamespace } from './PhpNamespace';
export { Printer } from './Printer';
export { PromotedParameter } from './PromotedParameter';
export { Property } from './Property';
export { PsrPrinter } from './PsrPrinter';
export { TraitType } from './TraitType';
export { TraitUse } from './TraitUse';
