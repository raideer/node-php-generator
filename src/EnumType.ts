import { HasConstants } from './Api/HasConstants';
import { HasMethods } from './Api/HasMethods';
import { HasTraits } from './Api/HasTraits';
import { ClassLike } from './ClassLike';
import { Constant } from './Constant';
import { EnumCase, EnumCaseValue } from './EnumCase';
import { Method } from './Method';
import { TraitUse } from './TraitUse';

export type EnumMember = Constant | Method | TraitUse | EnumCase;

export class EnumType
  extends ClassLike
  implements HasConstants, HasMethods, HasTraits
{
  private consts: Map<string, Constant> = new Map();
  private methods: Map<string, Method> = new Map();
  private traits: Map<string, TraitUse> = new Map();

  private implements: string[] = [];
  private cases: Map<string, EnumCase> = new Map();
  private type: string | undefined;

  public isEnum(): boolean {
    return true;
  }

  public setType(type?: string): this {
    this.type = type;
    return this;
  }

  public getType(): string | undefined {
    return this.type;
  }

  public setImplements(impl: string[]): this {
    this.validateNames(impl);
    this.implements = impl;
    return this;
  }

  public getImplements(): string[] {
    return this.implements;
  }

  public addImplement(impl: string): this {
    this.validateNames([impl]);
    this.implements.push(impl);
    return this;
  }

  public removeImplement(impl: string): this {
    this.implements = this.implements.filter((i) => i !== impl);
    return this;
  }

  public setCases(cases: EnumCase[]): this {
    this.cases.clear();
    for (const c of cases) {
      this.cases.set(c.getName(), c);
    }
    return this;
  }

  public getCases(): EnumCase[] {
    return Array.from(this.cases.values());
  }

  public addCase(
    name: string,
    value: EnumCaseValue = null,
    overwrite: boolean = false
  ): EnumCase {
    if (!overwrite && this.cases.has(name)) {
      throw new Error(`Case ${name} already exists`);
    }
    const c = new EnumCase(name);
    c.setValue(value);
    this.cases.set(name, c);
    return c;
  }

  public removeCase(name: string): this {
    this.cases.delete(name);
    return this;
  }

  public hasCase(name: string): boolean {
    return this.cases.has(name);
  }

  public addMember(member: EnumMember, overwrite: boolean = false): this {
    const name = member.getName();

    if (member instanceof Constant) {
      if (!overwrite && this.consts.has(name)) {
        throw new Error(`Constant ${name} already exists`);
      }
      this.consts.set(name, member);
      return this;
    } else if (member instanceof Method) {
      if (!overwrite && this.methods.has(name)) {
        throw new Error(`Method ${name} already exists`);
      }
      this.methods.set(name, member);
      return this;
    } else if (member instanceof TraitUse) {
      if (!overwrite && this.traits.has(name)) {
        throw new Error(`Trait ${name} already exists`);
      }
      this.traits.set(name, member);
      return this;
    } else if (member instanceof EnumCase) {
      if (!overwrite && this.cases.has(name)) {
        throw new Error(`Case ${name} already exists`);
      }
      this.cases.set(name, member);
      return this;
    }

    throw new Error('Invalid member');
  }

  public getConstants(): Constant[] {
    return Array.from(this.consts.values());
  }

  public getConstant(name: string): Constant | undefined {
    return this.consts.get(name);
  }

  public addConstant(
    name: string,
    value: string | undefined,
    override: boolean = false
  ): Constant {
    if (!override && this.consts.has(name)) {
      throw new Error(`Constant ${name} already exists`);
    }
    const constant = new Constant(name);
    constant.setValue(value);
    this.consts.set(name, constant);
    return constant;
  }

  public setConstants(constants: Constant[]): this {
    this.consts.clear();
    for (const constant of constants) {
      this.consts.set(constant.getName(), constant);
    }
    return this;
  }

  public removeConstant(name: string): this {
    this.consts.delete(name);
    return this;
  }

  public hasConstant(name: string): boolean {
    return this.consts.has(name);
  }

  public getMethods(): Method[] {
    return Array.from(this.methods.values());
  }

  public setMethods(methods: Method[]): this {
    this.methods.clear();
    for (const method of methods) {
      this.methods.set(method.getName(), method);
    }
    return this;
  }

  public getMethod(name: string): Method | undefined {
    return this.methods.get(name);
  }

  public addMethod(name: string, overwrite: boolean = false): Method {
    if (!overwrite && this.methods.has(name)) {
      throw new Error(`Method ${name} already exists`);
    }
    const method = new Method(name);
    this.methods.set(name, method);
    return method;
  }

  public removeMethod(name: string): this {
    this.methods.delete(name);
    return this;
  }

  public hasMethod(name: string): boolean {
    return this.methods.has(name);
  }

  public getTraits(): TraitUse[] {
    return Array.from(this.traits.values());
  }

  public setTraits(traits: TraitUse[]): this {
    this.traits.clear();
    for (const trait of traits) {
      this.traits.set(trait.getName(), trait);
    }
    return this;
  }

  public getTrait(name: string): TraitUse | undefined {
    return this.traits.get(name);
  }

  public addTrait(name: string, overwrite: boolean = false): TraitUse {
    if (!overwrite && this.traits.has(name)) {
      throw new Error(`Trait ${name} already exists`);
    }
    const trait = new TraitUse(name);
    this.traits.set(name, trait);
    return trait;
  }

  public removeTrait(name: string): this {
    this.traits.delete(name);
    return this;
  }

  public hasTrait(name: string): boolean {
    return this.traits.has(name);
  }

  public validate(): void {}
}
