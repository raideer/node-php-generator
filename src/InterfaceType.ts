import { ClassLike } from "./ClassLike";
import { Constant } from "./Constant";
import { Method } from "./Method";
import { withConstants } from "./mixins/withConstants";
import { withMethods } from "./mixins/withMethods";

export type InterfaceTypeMember = Constant | Method;

export class InterfaceType extends withConstants(withMethods(ClassLike)) {
  extends: string[] = [];

  getExtends(): string[] {
    return this.extends;
  }

  setExtends(interfaceExtends: string[]): this {
    this.extends = interfaceExtends;

    return this;
  }

  addExtends(interfaceExtends: string): this {
    this.extends.push(interfaceExtends);

    return this;
  }

  addMember(member: InterfaceTypeMember): this {
    const name = member.getName() as string;

    if (member instanceof Constant) {
      this.constants[name] = member;
    } else if (member instanceof Method) {
      this.methods[name] = member;
    }

    return this;
  }
}
