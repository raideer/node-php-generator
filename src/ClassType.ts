import { ClassLike } from "@/ClassLike";
import { PhpNamespace } from "@/PhpNamespace";
import { Constant } from "@/Constant";
import { Method } from "@/Method";
import { Property } from "@/Property";
import { withConstants } from "@/mixins/withConstants";
import { withMethods } from "@/mixins/withMethods";
import { withProperties } from "@/mixins/withProperties";
import { withTraits } from "@/mixins/withTraits";
import { TraitUse } from "@/TraitUse";

export type ClassTypeMember = Constant | Method | Property | TraitUse;

export class ClassType extends withConstants(withMethods(withProperties(withTraits(ClassLike)))) {
  private final = false;
  private abstract = false;
  private extends: string | null = null;
  private readOnly = false;
  private implements: string[] = [];

  isClass(): boolean {
    return true;
  }

  constructor(name: string, namespace: PhpNamespace | null = null) {
    super(name, namespace);
  }

  setFinal(final: boolean = true): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setAbstract(abstract: boolean = true): this {
    this.abstract = abstract;

    return this;
  }

  setReadOnly(readOnly: boolean = true): this {
    this.readOnly = readOnly;

    return this;
  }

  isReadOnly(): boolean {
    return this.readOnly;
  }

  isAbstract(): boolean {
    return this.abstract;
  }

  setExtends(ext: string | null): this {
    this.extends = ext;

    return this;
  }

  getExtends(): string | null {
    return this.extends;
  }

  setImplements(impl: string[]): this {
    this.implements = impl;

    return this;
  }

  getImplements(): string[] {
    return this.implements;
  }

  addMember(member: ClassTypeMember, override: boolean = false): this {
    const name = member.getName() as string;

    if (member instanceof Constant) {
      this.constants[name] = member;
    } else if (member instanceof Method) {
      this.methods[name] = member;
    } else if (member instanceof Property) {
      this.properties[name] = member;
    } else if (member instanceof TraitUse) {
      this.traits[name] = member;
    }

    return this;
  }
}
