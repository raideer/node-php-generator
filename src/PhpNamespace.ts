import { ClassType } from './ClassType';
import { InterfaceType } from './InterfaceType';
import { TraitType } from './TraitType';
import { EnumType } from './EnumType';
import { GlobalFunction } from './GlobalFunction';
import { Helpers } from './Helpers';

export enum NamespaceType {
  Normal = 'normal',
  Function = 'function',
  Constant = 'constant',
}

export class PhpNamespace {
  private name: string;
  private bracketedSyntax: boolean = false;
  private aliases: Record<string, Record<string, string>> = {
    [NamespaceType.Normal]: {},
    [NamespaceType.Function]: {},
    [NamespaceType.Constant]: {},
  };
  private classes: Record<
    string,
    ClassType | InterfaceType | TraitType | EnumType
  > = {};
  private functions: Record<string, GlobalFunction> = {};

  constructor(name: string) {
    if (name !== '' && !Helpers.isNamespaceIdentifier(name)) {
      throw new Error(`Value '${name}' is not valid name.`);
    }
    this.name = name;
  }

  public getName(): string {
    return this.name;
  }

  public setBracketedSyntax(state: boolean = true): this {
    this.bracketedSyntax = state;
    return this;
  }

  public hasBracketedSyntax(): boolean {
    return this.bracketedSyntax;
  }

  public addUse(
    name: string,
    alias: string | null = null,
    of: string = NamespaceType.Normal
  ): this {
    if (
      !Helpers.isNamespaceIdentifier(name, true) ||
      (Helpers.isIdentifier(name) && Helpers.Keywords[name.toLowerCase()])
    ) {
      throw new Error(
        `Value '${name}' is not valid class/function/constant name.`
      );
    } else if (
      alias &&
      (!Helpers.isIdentifier(alias) || Helpers.Keywords[alias.toLowerCase()])
    ) {
      throw new Error(`Value '${alias}' is not valid alias.`);
    }

    name = name.replace(/^\\/, '');
    const aliases = this.aliases[of];
    const used = of === NamespaceType.Normal ? this.classes : this.functions;

    if (alias === null) {
      let base = Helpers.extractShortName(name);
      let counter = 0;
      do {
        alias = base + (counter ? counter : '');
        counter++;
      } while (
        aliases[alias.toLowerCase()] &&
        aliases[alias.toLowerCase()] !== name
      );
    } else {
      if (
        aliases[alias.toLowerCase()] &&
        aliases[alias.toLowerCase()] !== name
      ) {
        throw new Error(
          `Alias '${alias}' used already for '${aliases[alias.toLowerCase()]}', cannot use for '${name}'.`
        );
      } else if (used[alias.toLowerCase()]) {
        throw new Error(
          `Name '${alias}' used already for '${this.name}\\${used[alias.toLowerCase()].getName()}'.`
        );
      }
    }

    aliases[alias] = name;
    return this;
  }

  public removeUse(name: string, of: string = NamespaceType.Normal): void {
    for (const alias in this.aliases[of]) {
      if (this.aliases[of][alias].toLowerCase() === name.toLowerCase()) {
        delete this.aliases[of][alias];
      }
    }
  }

  public addUseFunction(name: string, alias: string | null = null): this {
    return this.addUse(name, alias, NamespaceType.Function);
  }

  public addUseConstant(name: string, alias: string | null = null): this {
    return this.addUse(name, alias, NamespaceType.Constant);
  }

  public getUses(of: string = NamespaceType.Normal): Record<string, string> {
    const aliases = this.aliases[of];

    // Sort the aliases based on the transformed string
    const sortedAliases = Object.entries(aliases).sort(
      ([aliasA, nameA], [aliasB, nameB]) => {
        const transformedA = nameA.replace(/\\/g, ' ');
        const transformedB = nameB.replace(/\\/g, ' ');
        return transformedA.localeCompare(transformedB);
      }
    );

    // Filter the aliases
    const filteredAliases = sortedAliases.filter(([alias, name]) => {
      const fullAlias = (this.name ? this.name + '\\' : '') + alias;
      return (
        fullAlias.localeCompare(name, undefined, { sensitivity: 'base' }) !== 0
      );
    });

    return Object.fromEntries(filteredAliases);
  }

  public resolveName(name: string, of: string = NamespaceType.Normal): string {
    if (Helpers.Keywords[name.toLowerCase()] || name === '') {
      return name;
    } else if (name[0] === '\\') {
      return name.substring(1);
    }

    const aliases = this.aliases[of];
    if (of !== NamespaceType.Normal) {
      return (
        aliases[name.toLowerCase()] ||
        this.resolveName(Helpers.extractNamespace(name) + '\\') +
          Helpers.extractShortName(name)
      );
    }

    const parts = name.split('\\', 2);
    return (
      aliases[parts[0].toLowerCase()] ||
      this.name + (this.name ? '\\' : '') + name
    );
  }

  public simplifyType(type: string, of: string = NamespaceType.Normal): string {
    return type.replace(/[\w\x7f-\xff\\]+/g, (match) =>
      this.simplifyName(match, of)
    );
  }

  public simplifyName(name: string, of: string = NamespaceType.Normal): string {
    if (Helpers.Keywords[name.toLowerCase()] || name === '') {
      return name;
    }

    name = name.replace(/^\\/, '');

    if (of !== NamespaceType.Normal) {
      for (const alias in this.aliases[of]) {
        if (this.aliases[of][alias].toLowerCase() === name.toLowerCase()) {
          return alias;
        }
      }

      return (
        this.simplifyName(Helpers.extractNamespace(name) + '\\') +
        Helpers.extractShortName(name)
      );
    }

    let shortest: string | null = null;
    let relative = name.startsWith(this.name + '\\')
      ? name.substring(this.name.length + 1)
      : null;

    for (const alias in this.aliases[of]) {
      const original = this.aliases[of][alias];
      if (relative && (relative + '\\').startsWith(alias + '\\')) {
        relative = null;
      }

      if ((name + '\\').startsWith(original + '\\')) {
        const short = alias + name.substring(original.length);
        if (!shortest || shortest.length > short.length) {
          shortest = short;
        }
      }
    }

    if (shortest && relative && shortest.length < relative.length) {
      return shortest;
    }

    return relative ?? shortest ?? (this.name ? '\\' : '') + name;
  }

  public add(
    classType: ClassType | InterfaceType | TraitType | EnumType
  ): this {
    const name = classType.getName();
    if (!name) {
      throw new Error('Class does not have a name.');
    }

    const lower = name.toLowerCase();
    if (this.classes[lower] && this.classes[lower] !== classType) {
      throw new Error(`Cannot add '${name}', because it already exists.`);
    } else if (this.aliases[NamespaceType.Normal][lower]) {
      throw new Error(
        `Name '${name}' used already as alias for ${this.aliases[NamespaceType.Normal][lower]}.`
      );
    }

    this.classes[lower] = classType;
    return this;
  }

  public addClass(name: string): ClassType {
    const classType = new ClassType(name, this);
    this.add(classType);
    return classType;
  }

  public addInterface(name: string): InterfaceType {
    const interfaceType = new InterfaceType(name, this);
    this.add(interfaceType);
    return interfaceType;
  }

  public addTrait(name: string): TraitType {
    const traitType = new TraitType(name, this);
    this.add(traitType);
    return traitType;
  }

  public addEnum(name: string): EnumType {
    const enumType = new EnumType(name, this);
    this.add(enumType);
    return enumType;
  }

  public getClass(
    name: string
  ): ClassType | InterfaceType | TraitType | EnumType {
    const classType = this.classes[name.toLowerCase()];
    if (!classType) {
      throw new Error(`Class '${name}' not found.`);
    }
    return classType;
  }

  public getClasses(): (ClassType | InterfaceType | TraitType | EnumType)[] {
    return Object.values(this.classes);
  }

  public removeClass(name: string): this {
    delete this.classes[name.toLowerCase()];
    return this;
  }

  public addFunction(name: string): GlobalFunction {
    const lower = name.toLowerCase();
    if (this.functions[lower]) {
      throw new Error(`Cannot add '${name}', because it already exists.`);
    } else if (this.aliases[NamespaceType.Function][lower]) {
      throw new Error(
        `Name '${name}' used already as alias for ${this.aliases[NamespaceType.Function][lower]}.`
      );
    }

    const func = new GlobalFunction(name);
    this.functions[lower] = func;
    return func;
  }

  public getFunction(name: string): GlobalFunction {
    const func = this.functions[name.toLowerCase()];
    if (!func) {
      throw new Error(`Function '${name}' not found.`);
    }
    return func;
  }

  public getFunctions(): GlobalFunction[] {
    return Object.values(this.functions);
  }

  public removeFunction(name: string): this {
    delete this.functions[name.toLowerCase()];
    return this;
  }
}
