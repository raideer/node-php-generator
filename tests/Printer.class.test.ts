import { Printer } from '../src/Printer';
import { ClassType } from '../src/ClassType';
import { InterfaceType } from '../src/InterfaceType';
import { TraitType } from '../src/TraitType';
import { EnumType } from '../src/EnumType';
import { Method } from '../src/Method';
import { Property } from '../src/Property';
import { PhpNamespace } from '../src/PhpNamespace';

test('prints basic class', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints abstract class', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.setAbstract();

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints final class', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.setFinal();

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints readonly class', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.setReadOnly();

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with extends', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.setExtends('BaseClass');

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with implements', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.setImplements(['Interface1', 'Interface2']);

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with properties', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');

  const prop1 = new Property('prop1');
  prop1.setPublic();
  prop1.setType('string');

  const prop2 = new Property('prop2');
  prop2.setPrivate();
  prop2.setType('int');
  prop2.setValue(42);

  classType.setProperties([prop1, prop2]);

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with methods', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');

  const method1 = new Method('method1');
  method1.setPublic();
  method1.setReturnType('string');

  const method2 = new Method('method2');
  method2.setPrivate();
  method2.addParameter('param');

  classType.setMethods([method1, method2]);

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with traits', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.addTrait('Trait1');
  classType.addTrait('Trait2');

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints class with attributes', () => {
  const printer = new Printer();
  const classType = new ClassType('Foo');
  classType.addAttribute('Entity', { table: 'foo_table' });

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});

test('prints interface', () => {
  const printer = new Printer();
  const interfaceType = new InterfaceType('IFoo');

  const method = new Method('doSomething');
  method.setReturnType('void');
  interfaceType.setMethods([method]);

  const output = printer.printClass(interfaceType);

  expect(output).toMatchSnapshot();
});

test('prints interface with extends', () => {
  const printer = new Printer();
  const interfaceType = new InterfaceType('IFoo');
  interfaceType.setExtends(['IBase1', 'IBase2']);

  const output = printer.printClass(interfaceType);

  expect(output).toMatchSnapshot();
});

test('prints trait', () => {
  const printer = new Printer();
  const traitType = new TraitType('MyTrait');

  const property = new Property('prop');
  property.setPrivate();

  const method = new Method('traitMethod');
  method.setPublic();

  traitType.setProperties([property]);
  traitType.setMethods([method]);

  const output = printer.printClass(traitType);

  expect(output).toMatchSnapshot();
});

test('prints enum', () => {
  const printer = new Printer();
  const enumType = new EnumType('Status');

  enumType.addCase('PENDING');
  enumType.addCase('ACTIVE', 1);
  enumType.addCase('INACTIVE', 2);

  const output = printer.printClass(enumType);

  expect(output).toMatchSnapshot();
});

test('prints backed enum', () => {
  const printer = new Printer();
  const enumType = new EnumType('Status');
  enumType.setType('string');

  enumType.addCase('PENDING', 'pending');
  enumType.addCase('ACTIVE', 'active');

  const output = printer.printClass(enumType);

  expect(output).toMatchSnapshot();
});

test('prints class with namespace resolution', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App\\Entity');
  const classType = new ClassType('User');

  classType.setExtends('BaseEntity');
  classType.setImplements(['IdentifiableInterface']);
  namespace.addUse('App\\Base\\BaseEntity');
  namespace.addUse('App\\Contracts\\IdentifiableInterface');

  const output = printer.printClass(classType, namespace);

  expect(output).toMatchSnapshot();
});

test('prints complex class', () => {
  const printer = new Printer();
  const classType = new ClassType('ComplexClass');

  // Class modifiers
  classType.setFinal();
  classType.setReadOnly();

  // Inheritance
  classType.setExtends('BaseClass');
  classType.setImplements(['Interface1', 'Interface2']);
  classType.addTrait('Trait1');

  // Properties
  const prop1 = new Property('prop1');
  prop1.setPublic();
  prop1.setReadOnly();
  prop1.setType('string');

  const prop2 = new Property('prop2');
  prop2.setPrivate();
  prop2.setStatic();
  prop2.setValue(['a', 'b', 'c']);

  classType.setProperties([prop1, prop2]);

  // Methods
  const method1 = new Method('method1');
  method1.setPublic();
  method1.setReturnType('array');
  method1.addParameter('param').setType('string');

  const method2 = new Method('method2');
  method2.setProtected();
  method2.setStatic();
  method2.setReturnType('void');

  classType.setMethods([method1, method2]);

  // Attributes
  classType.addAttribute('Entity', { table: 'complex' });
  classType.addAttribute('Cached');

  const output = printer.printClass(classType);

  expect(output).toMatchSnapshot();
});
