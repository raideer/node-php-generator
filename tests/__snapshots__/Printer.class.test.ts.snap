// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints abstract class 1`] = `
"abstract class Foo
{
}
"
`;

exports[`prints backed enum 1`] = `
"enum Status: string
{
    case PENDING = 'pending';
    case ACTIVE = 'active';
}
"
`;

exports[`prints basic class 1`] = `
"class Foo
{
}
"
`;

exports[`prints class with attributes 1`] = `
"#[Entity(table: 'foo_table')]
class Foo
{
}
"
`;

exports[`prints class with extends 1`] = `
"class Foo extends BaseClass
{
}
"
`;

exports[`prints class with implements 1`] = `
"class Foo implements Interface1, Interface2
{
}
"
`;

exports[`prints class with methods 1`] = `
"class Foo
{
    public function method1(): string
    {
    }

    private function method2($param)
    {
    }
}
"
`;

exports[`prints class with namespace resolution 1`] = `
"class User extends \\BaseEntity implements \\IdentifiableInterface
{
}
"
`;

exports[`prints class with properties 1`] = `
"class Foo
{
    public string $prop1 = null;
    private int $prop2 = 42;
}
"
`;

exports[`prints class with traits 1`] = `
"class Foo
{
    use Trait1;
    use Trait2;
}
"
`;

exports[`prints complex class 1`] = `
"#[Entity(table: 'complex')]
#[Cached]
final readonly class ComplexClass extends BaseClass implements Interface1, Interface2
{
    use Trait1;

    public string $prop1 = null;
    private static $prop2 = [0 => 'a', 1 => 'b', 2 => 'c'];

    public function method1(string $param): array
    {
    }

    protected static function method2(): void
    {
    }
}
"
`;

exports[`prints enum 1`] = `
"enum Status
{
    case PENDING;
    case ACTIVE = 1;
    case INACTIVE = 2;
}
"
`;

exports[`prints final class 1`] = `
"final class Foo
{
}
"
`;

exports[`prints interface 1`] = `
"interface IFoo
{
    function doSomething(): void;
}
"
`;

exports[`prints interface with extends 1`] = `
"interface IFoo extends IBase1, IBase2
{
}
"
`;

exports[`prints readonly class 1`] = `
"readonly class Foo
{
}
"
`;

exports[`prints trait 1`] = `
"trait MyTrait
{
    private $prop = null;

    public function traitMethod()
    {
    }
}
"
`;
